pkg/probe/system_vdom_resources.go

```diff
@@ -0,0 +1,120 @@
+package probe
+
+import (
+	"encoding/json"
+	"log"
+	"strconv"
+
+	"github.com/bluecmd/fortigate_exporter/pkg/http"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type HAVDOMResults struct {
+	ID                 float64 `json:"id"`
+	CustomMax          int     `json:"custom_max"`
+	MinCustomValue     int     `json:"min_custom_value"`
+	MaxCustomValue     int     `json:"max_custom_value"`
+	Guaranteed         int     `json:"guaranteed"`
+	MinGuaranteedValue int     `json:"min_guaranteed_value"`
+	MaxGuaranteedValue int     `json:"max_guaranteed_value"`
+	GlobalMax          int     `json:"global_max"`
+	CurrentUsage       int     `json:"current_usage"`
+	UsagePercent       int     `json:"usage_percent"`
+}
+
+type HAResults struct {
+	Cpu         float64 `json:"cpu"`
+	Memory      float64 `json:"memory"`
+	SetupRate   float64 `json:"setup_rate"`
+	IsDeletable bool    `json:"is_deletable"`
+	Interfaces  map[string]HAVDOMResults
+}
+
+func (o *HAResults) SystemVDOMResourcesUnmarshalJSON(data []byte) error {
+	var m map[string]json.RawMessage
+	if err := json.Unmarshal(data, &m); err != nil {
+		return err
+	}
+
+	if _, found := m["cpu"]; found {
+		if err := json.Unmarshal(m["cpu"], &o.Cpu); err != nil {
+			return err
+		}
+		delete(m, "cpu")
+	}
+
+	if _, found := m["memory"]; found {
+		if err := json.Unmarshal(m["memory"], &o.Memory); err != nil {
+			return err
+		}
+		delete(m, "memory")
+	}
+
+	if _, found := m["setup_rate"]; found {
+		if err := json.Unmarshal(m["setup_rate"], &o.SetupRate); err != nil {
+			return err
+		}
+		delete(m, "setup_rate")
+	}
+
+	if _, found := m["is_deletable"]; found {
+		if err := json.Unmarshal(m["is_deletable"], &o.IsDeletable); err != nil {
+			return err
+		}
+		delete(m, "is_deletable")
+	}
+
+	o.Interfaces = make(map[string]HAVDOMResults)
+	for k, v := range m {
+		var p HAVDOMResults
+		if err := json.Unmarshal(v, &p); err != nil {
+			return err
+		}
+		o.Interfaces[k] = p
+	}
+	return nil
+}
+
+func probeSystemVDOMResourcesAll(c http.FortiHTTP, meta *TargetMetadata) ([]prometheus.Metric, bool) {
+	var (
+		vdomResources = prometheus.NewDesc(
+			"fortigate_vdom_resources",
```

Comment:
These metrics need to be split up, one per metric. Having aggregated/"catch-all" ones like this makes little sense. See https://prometheus.io/docs/practices/naming/#metric-names, especially the last part: "...should represent the same logical thing-being-measured across all label dimensions. ".

Also:

> As a rule of thumb, either the sum() or the avg() over all dimensions of a given metric should be meaningful (though not necessarily useful). 

Doing `sum(fortigate_vdom_resources)` seems to me to not produce anything that makes any sense. Sticking values into the labels will explode cardinality as well as make it non-ideal to access the value for graphing purposes.

Finally, I cannot tell what the metrics are actually measuring from the name. E.g. what is `fortigate_vdom_resources{usage_percent}` or `fortigate_vdom_resources{custom_max}`?

Reply:
> These metrics need to be split up, one per metric. Having aggregated/"catch-all" ones like this makes little sense. See https://prometheus.io/docs/practices/naming/#metric-names, especially the last part: "...should represent the same logical thing-being-measured across all label dimensions. ".
> 
> Also:
> 
> > As a rule of thumb, either the sum() or the avg() over all dimensions of a given metric should be meaningful (though not necessarily useful).
> 
> Doing `sum(fortigate_vdom_resources)` seems to me to not produce anything that makes any sense. Sticking values into the labels will explode cardinality as well as make it non-ideal to access the value for graphing purposes.
> 
> Finally, I cannot tell what the metrics are actually measuring from the name. E.g. what is `fortigate_vdom_resources{usage_percent}` or `fortigate_vdom_resources{custom_max}`?

I haven't yet verified what `usagepercent` is derieved from. I will have a look and get back on it but the practical idea here is if you let's say you have a big 3500 cluster for example and you have a maximum of let's say 35 million concurrent sessions. You'd ideally want to allocate sessions to each VDOM. 

Take in this example you'd allocate it to the Guaranteed number of sesisons, if you then want to make a Prometheus alert on it. You'd just divide the current usage by the Guaranteed to get the usage percentage. As it stands currently without this value you have to make a custom alert rule for each defined limit.